middle_end/flambda2/terms/flambda_primitive.ml

```diff
@@ -995,7 +995,11 @@ let effects_and_coeffects_of_unary_primitive p : Effects_and_coeffects.t =
     (* [Obj.truncate] has now been removed. *)
     No_effects, No_coeffects, Strict
   | String_length _ -> No_effects, No_coeffects, Strict
-  | Int_as_pointer -> No_effects, No_coeffects, Strict
+  | Int_as_pointer alloc_mode ->
+    let coeffects : Coeffects.t =
+      match alloc_mode with Local _ -> Has_coeffects | Heap -> No_coeffects
+    in
```

Comment:
You should lift this out to it's own function and also use it for e.g. `Box_number`.